File: Leetcode/0310-MinimumHeightTrees/Solution.java
class Solution {
    public List<Integer> findMinHeightTrees(int n, int[][] edges) {
        // Edge case
        if (n <= 2) {
            List<Integer> result = new ArrayList<>();
            for (int v = 0; v < n; v++) {
                result.add(v);
            }
            return result;
        }
        // Build undirected graph
        Set<Integer>[] graph = new HashSet[n];
        for (int v = 0; v < n; v++) {
            graph[v] = new HashSet<>();
        }
        // Add edges
        for (int[] edge : edges) {
            graph[edge[0]].add(edge[1]);
            graph[edge[1]].add(edge[0]);
        }
        // BFS
        Queue<Integer> q = new LinkedList<>();
        // Add leaves to queue
        for (int v = 0; v < n; v++) {
            if (graph[v].size() == 1) {
                q.add(v);
            }
        }
        // n vertice remaining
        int remain = n;
        while (remain > 2) {
            // Each time dequeue all items
            int size = q.size();
            remain -= size;
            while (size > 0) {
                int leaf = q.remove();
                // Remove leaf from all its neighbor's edges
                for (int neighbor : graph[leaf]) {
                    graph[neighbor].remove(leaf);
                    // New leaf
                    if (graph[neighbor].size() == 1) {
                        q.add(neighbor);
                    }
                }
                size--;
            }
        }
        List<Integer> list = new ArrayList();
        list.addAll(q);
        return list;
    }}

    

    

File: Leetcode/0310-MinimumHeightTrees/solution.cpp
#include <queue>
#include <vector>

using namespace std;

class Solution {
 public:
  vector<int> findMinHeightTrees(int n, vector<vector<int>>& edges) {
    // Edge case: when there's only 1 node, its indegree is 0
    if (n == 1) {
      return {0};
    }
    // Build graph
    vector<vector<int>> graph(n);
    vector<int> indegrees(n, 0);
    for (const vector<int>& edge : edges) {
      graph[edge[0]].push_back(edge[1]);
      graph[edge[1]].push_back(edge[0]);
      indegrees[edge[0]]++;
      indegrees[edge[1]]++;
    }
    // Similar to topological sorting, start visiting from nodes with 1 indegree
    queue<int> q;
    for (int u = 0; u < n; u++) {
      if (indegrees[u] == 1) {
        // Push leaf node to queue
        q.push(u);
      }
    }
    int unvisited = n;  // number of unvisited nodes
    while (unvisited > 2) {
      int size = q.size();
      while (size > 0) {
        int u = q.front();
        q.pop();
        unvisited--;
        // Remove indegrees related to u
        indegrees[u] = 0;
        for (int v : graph[u]) {
          indegrees[v]--;
          if (indegrees[v] == 1) {
            // Push the next leaf node to queue
            q.push(v);
          }
        }
        size--;
      }
    }
    // Return result
    vector<int> res;
    while (!q.empty()) {
      res.push_back(q.front());
      q.pop();
    }
    return res;
  }
};
